### PR TITLE
Add Beok TOL47 wifi thermostat to supported devices

### PR DESCRIPTION
**Pull request by kekkutya**
https://www.beok-controls.com/Content/upload/PDF/2021837153/TOL47-Heating-Thermostat-Manual.pdf

### CODE REVIEW

File: custom_components/tuya_local/devices/beok_tol47_thermostat.yaml [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+          step: 5
+    - id: 3
+      type: integer
+      name: current_temperature
+      mapping:
+        - scale: 10
+    - id: 4
+      type: string
+      name: preset_mode
+      mapping:
```
Comment (make-all) [inline custom_components/tuya_local/devices/beok_tol47_thermostat.yaml:26]: This seems like a strange mapping.  Are you sure you got preset_mode and hvac_mode around the right way?

File: custom_components/tuya_local/devices/beok_tol47_thermostat.yaml [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+          - dps_val: 3
+            icon: "mdi:lightbulb-on"
+            value: High
+  - entity: select
+    name: Schedule
+    category: config
+    icon: "mdi:calendar-clock"
+    dps:
+      - id: 42
+        type: integer
```
Comment (make-all) [inline custom_components/tuya_local/devices/beok_tol47_thermostat.yaml:150]: Normally enum types in tuya are strings, even if they use numbers as the value (they will have quotes around them in the raw Tuya protocol).  This integration is lenient with string/integer distinction, so it probably works either way but better for the config to use the correct one.

File: custom_components/tuya_local/devices/beok_tol47_thermostat.yaml (hunk 1 above)
Comment (kekkutya) [reply custom_components/tuya_local/devices/beok_tol47_thermostat.yaml:26]: Yes, a little bit strange for me too. Debugged dps are:

```
"device_config": {
"friendly_name": "Living Room Thermostat",
"local_key": "xxxxxxxxxxxxxxx",
"host": "192.168.0.55",
"device_id": "xxxxxxxxxxxxxxx",
"protocol_version": "3.3",
"scan_interval": 1,
"dps_strings": [
	"1 (value: True)",	# Switch
	"2 (value: 205)",	# Set Temperature *10
	"3 (value: 202)", 	# Current Temperature *10
	"4 (value: cold)",	# Mode "cold": Manual "hot": Program "wind": Temporary Program 
	"5 (value: 1)",		# Working status 0: off 1: on
	"8 (value: False)",	# Child lock
	"13 (value: True)",	# Sound
	"16 (value: 0)",	# Fault alarm
	"20 (value: -7)",	# Temperature calibration *10
	"21 (value: 30)",	# Max setting temperature
	"25 (value: 1)",        # Sensor selection 1: internal 2: external 3: both
	"26 (value: False)",    # Frost protection
	"31 (value: False)",	# Reset
	"41 (value: 3)",       	# Backlight brightness 0: Close 1: Low 2: Middle 3: High
	"42 (value: 1)",      	# Working day settings 1: 5+2 2: 6+1 3: 7
 	"101 (value: False)",	# Output main reverse
 	"105 (value: 5)",     	# Temperature control switch difference *10
 	"107 (value: 30)"       # External Sensor temperature limit
```
- on dp4, i can change value with the Tuya Smart app "Mode" settings, and if i change it from HA, the app settings reflect correctly.
- on dp5, value 0 if dry contact is open, 1 if closed (=boyler heating)
        		
Or i completely wrong interpreting preset_mode and hvac_mode?

File: custom_components/tuya_local/devices/beok_tol47_thermostat.yaml (hunk 2 above)
Comment (kekkutya) [reply custom_components/tuya_local/devices/beok_tol47_thermostat.yaml]: Thanks for your comment (very beginner with this), modified variable types (+ add dp16 Fault alarm settings)

File: custom_components/tuya_local/devices/beok_tol47_thermostat.yaml (hunk 1 above)
Comment (make-all) [reply custom_components/tuya_local/devices/beok_tol47_thermostat.yaml:26]: In that case, preset_mode is correct for dp 4, but I would use hvac_action for dp 5, and hvac_mode directly for dp 1.

**Comment by kekkutya** [comment]
> I'll fix up the hvac_action and hvac_mode later

Thank you,
i finished at now.

feel free to use it or modify. Thanks for the revision and help (it was my first activity on github)
  
name: Beok TOL47 thermostat
primary_entity:
  entity: climate
  dps:
    - id: 1
      type: boolean
      name: hvac_mode
      mapping:
        - dps_val: true
          value: "heat"
        - dps_val: false
          value: "off"
    - id: 2
      type: integer
      name: temperature
      mapping:
        - scale: 10
          step: 5
    - id: 3
      type: integer
      name: current_temperature
      mapping:
        - scale: 10
    - id: 4
      type: string
      name: preset_mode
      mapping:
        - dps_val: "cold"
          value: Manual
        - dps_val: "hot"
          value: Program
        - dps_val: "wind"
          value: Temporary
    - id: 5
      type: string
      name: hvac_action
      mapping:
        - dps_val: "0"
          icon: "mdi:radiator-disabled"
          constraint: hvac_mode
          conditions:
            - dps_val: true
              value: idle
            - dps_val: false
              value: off
        - dps_val: "1"
          value: heating
          icon: "mdi:radiator"

Regards, Zoli

**Comment by make-all** [comment]
Sorry, it seems I did not quite wait long enough.  Was there anything else you changed  besides the hvac_mode and hvac_action I suggested above?

**Comment by escoand** [comment]
@kekkutya Do you know what this sound switch is useful for? I have the same device and always wonder if this could be used for something smart.

**Comment by make-all** [comment]
My understanding is that the wall panel emits a beep if you press a button, and this switch can be used to turn that on and off.

**Comment by kekkutya** [comment]
> My understanding is that the wall panel emits a beep if you press a button, and this switch can be used to turn that on and off. I guess one smart thing you could do is turn it off at night and on again in the morning with automations.

seems logical, but my device does not make sound at any position of the switch 
@make-all: may you check issue #289 by escoand pls.? Seems some incompatibility with awatto wt100, when add a new device.

**Comment by make-all** [comment]
A fix for #289 was pushed about an hour ago.  There was a syntax error in the config for Avatto WT100, which was not picked up by unit tests.

**Comment by escoand** [comment]
> seems logical, but my device does not make sound at any position of the switch

That's why I was asking...

**Comment by make-all** [comment]
This thermostat seems to share a lot in common with the Avatto WT100.  There may be a wider range by the same OEM manufacturer, with slight differences in spec.  It is possible that the beep hardware was left off the Beok TOL47, even though it is still exposed in the firmware.  It was mentioned in the Avatto manual that I read, and the original config came from the iot.tuya.com info for this device I suppose.  Does the official app expose this switch?  Maybe it is best moved to an attribute on the climate entity where it can still be observed, but the switch removed to prevent confusion.

**Comment by kekkutya** [comment]
> Does the official app expose this switch? 

Yes. As i see, the app screens exactly the same for both devices. And Beok's (not implemented yet) dp31 boolean is the "Reset" like in Avatto.

**Pull request merged**